### PR TITLE
python module: make it work with pypy

### PR DESCRIPTION
**Pull request by MathieuDuponchelle**
pypy installations don't usuallyy ship with pkg-config files,
we thus need to replicate what their version of distutils does.

In addition, we also try our best to build against other
pythons that do not have pkg-config files.

### CODE REVIEW

File: mesonbuild/modules/python.py [hunk 1]
```diff
@@ -238,6 +278,7 @@ def __init__(self, interpreter, python):
         self.purelib_install_path = os.path.join(prefix, install_paths['purelib'][1:])
         self.version = run_command(python, "import sysconfig; print (sysconfig.get_python_version())")
         self.platform = run_command(python, "import sysconfig; print (sysconfig.get_platform())")
+        self.is_pypy = json.loads(run_command (python, IS_PYPY_COMMAND))
```
Comment (ignatenkobrain) [inline mesonbuild/modules/python.py]: __[Flake8]__

[E211] whitespace before '('

File: mesonbuild/modules/python.py [hunk 2]
```diff
@@ -94,14 +94,46 @@ def __init__(self, python_holder, environment, kwargs):
...
+            libname = 'python{}'.format(self.version)
+            if 'DEBUG_EXT' in self.variables:
+                libname += self.variables['DEBUG_EXT']
+            if 'ABIFLAGS' in self.variables:
+                libname += self.variables['ABIFLAGS']
+            libdirs = []
+
+        largs = self.compiler.find_library(libname, environment, libdirs)
+
+        self.is_found = largs != None
```
Comment (ignatenkobrain) [inline mesonbuild/modules/python.py]: __[Flake8]__

[E711] comparison to None should be 'if cond is not None:'

File: mesonbuild/modules/python.py (hunk 1 above)
Comment (MathieuDuponchelle) [reply mesonbuild/modules/python.py]: That's fixed, not sure why it doesn't show as outdated :)

**Comment by MathieuDuponchelle** [comment]
Probably yes, for what it's worth pygobject uses pyenv to manage multiple pythons, see <https://gitlab.gnome.org/GNOME/pygobject/tree/master/.gitlab-ci> and <https://gitlab.gnome.org/GNOME/pygobject/blob/master/.gitlab-ci.yml>

**Comment by MathieuDuponchelle** [comment]
@jpakkane can we merge this?

**Comment by jpakkane** [comment]
I'd really like to get PyPy in the CI image so this does not get broken in the future. Is it enough for me install PyPy dev packages? Which ones does it need? Does the code need changing so the PyPy test is run as part of the suite?

**Comment by jpakkane** [comment]
I have started a CI image job that adds pypy-dev to the list of packages. That is version 5.8.0 which has only Python 2.7, though. Will update this comment once it has been uploaded.

**Comment by MathieuDuponchelle** [comment]
> That is version 5.8.0 which has only Python 2.7, though. Will update this comment once it has been uploaded.

ack, you'll probably want to restart the travis job once it is uploaded then :)

**Comment by jpakkane** [comment]
It seemed to already be green. But restarted anyway. Need to remember to check the log that it was actually run...

**Comment by jpakkane** [comment]
Apparently the `python` test dir is not run at all. Only `python3` is. The directory should be added to `all_tests` in `detect_tests_to_run`.

**Comment by MathieuDuponchelle** [comment]
@jpakkane it is run by run_unittests.py

**Comment by MathieuDuponchelle** [comment]
> In that case it should be under test cases/unit. 

OK, not really related to this issue though :)

**Comment by nirbheek** [comment]
On my system, if pypy is installed but the -devel package isn't installed, configure passes but the build fails.

**Comment by nirbheek** [comment]
I've renamed the test directory. It seems to me that with pypy, since a pkg-config file is not being used for detection, we should try to do a `cc.has_header('Python.h')` with the `c_args:` that are meant to be used for pypy.

**Comment by MathieuDuponchelle** [comment]
> we should try to do a cc.has_header('Python.h') with the c_args

Yes, that would probably be helpful :) re unit test, I think I'll split it up into individual test cases at some point, what gets skipped / succeeds / fails will be clearer that way

**Pull request merged**